### PR TITLE
Android integration

### PR DESCRIPTION
**Pull request by AgneseBussone**
Javascript code used by android copied by the local version of the renderer.

### CODE REVIEW

**Review by mortenbo — APPROVED** [review]

**Comment by mortenbo** [comment]
I just approved it but can we make it less offer specific? Incito is not offer specific but based on views.

**Comment by AgneseBussone** [comment]
@mortenbo  are you referring to `window.goToOffer` function?
We could change the name, but it will still look for an offer `incito__view[data-role="offer"]`, so it's kind of offer related.

**Comment by mortenbo** [comment]
I'd do something like this:

```javascript
window.goToView = function (viewId) {
    var all = document.getElementsByTagName("*");

    for (var i=0, max=all.length; i < max; i++) {
        if (all[i].webkitMatchesSelector('.incito__view') && all[i].getAttribute('data-id') === viewId) {
            all[i].scrollIntoView();

            break;
        }
    }
};
```

Also, the break is super important if you want performance.

File: index-1.0.0.html [hunk 1]
```diff
@@ -44,34 +48,69 @@
             });
 
             if ('androidJSInterface' in window) {
-
+                // Listen for click events, and forward the id of the offer to Android listeners.
                 document.addEventListener("click", function (e) {
-                    var id = findViewId(e.target);
+                    var id = viewIdOfFirstParentOffer(e.target);
```
Comment (mortenbo) [inline index-1.0.0.html]: We shouldn't tie us into offers in this repo, IMO.

File: index-1.0.0.html [hunk 2]
```diff
@@ -90,7 +129,17 @@
             };
         }
 
-        // Polyfill for `elementsFromPoint` for older versions of ios Safari
+        // internal function to retrieve the viewId of the first 'offer' parent of an element.
+        // Called from JS code triggered by the android java code.
+        function viewIdOfFirstParentOffer(element) {
+            for (var target = element; target && target != this; target = target.parentNode) {
+                if (target.webkitMatchesSelector('.incito__view[data-role="offer"]')) {
+                    return target.getAttribute('data-id');
```
Comment (mortenbo) [inline index-1.0.0.html]: Same here.

File: index-1.0.0.html (hunk 1 above)
Comment (lozhuf) [reply index-1.0.0.html]: agreed. discussing alternatives with @AgneseBussone now.

File: index-1.0.0.html [hunk 3]
```diff
@@ -28,33 +28,89 @@
...
-        //- Returns array of (non-null) id strings for all the elements at the specified point
+        // Returns array of (non-null) id strings for all the elements at the specified point
+        // [!] FUNCTION SIGNATURE MUST NOT CHANGE - Used in production
         function getElementIdsAtPoint(x, y) {
-            return [...document.elementsFromPoint(x, y)]
-                .map(el => el.getAttribute('data-id'))
-                .filter(el => el != null)
+            return document.elementsFromPoint(x, y)
+                .map(function (el) { return el.getAttribute('data-id') })
+                .filter(function (el) { return el != null })
```
Comment (klarstrup) [inline index-1.0.0.html:81]: You can use `reduce` here to go from O(2n) to O(n) 😁
```js
  return document.elementsFromPoint(x, y).reduce(function(els, el) {
    const dataId = el.getAttribute("data-id");
    if (dataId) els.push(dataId);
    return els;
  }, []);
```

Comment (lozhuf) [reply index-1.0.0.html:81]: I like it. I was hoping for a JS version of compactMap, but alas, reduce it is.

Comment (lozhuf) [reply index-1.0.0.html:81]: not as readable, tho, but that's reduce for you.

Comment (klarstrup) [reply index-1.0.0.html:81]: hey be nice to `reduce` it's gotten me through many tough times

Comment (lozhuf) [reply index-1.0.0.html:81]: flatMap is my jam

Comment (klarstrup) [reply index-1.0.0.html:81]: `flatMap` actually landed last year https://caniuse.com/#feat=array-flat we got dat

**Review by lozhuf — APPROVED** [review]

**Pull request merged**